File: ganttproject/src/net/sourceforge/ganttproject/DialogBuilder.java
/*
GanttProject is an opensource project management tool. License: GPL3
Copyright (C) 2012 GanttProject Team

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 3
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
package net.sourceforge.ganttproject;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.GridLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import javax.swing.AbstractAction;
import javax.swing.Action;
import javax.swing.BorderFactory;
import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JDialog;
import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.KeyStroke;
import javax.swing.SwingUtilities;
import javax.swing.WindowConstants;

import net.java.balloontip.BalloonTip;
import net.java.balloontip.styles.EdgedBalloonStyle;
import net.sourceforge.ganttproject.action.CancelAction;
import net.sourceforge.ganttproject.action.OkAction;
import net.sourceforge.ganttproject.gui.DialogAligner;
import net.sourceforge.ganttproject.gui.NotificationComponent.AnimationView;
import net.sourceforge.ganttproject.gui.NotificationManager;
import net.sourceforge.ganttproject.gui.UIFacade.Centering;
import net.sourceforge.ganttproject.gui.UIFacade.Dialog;
import net.sourceforge.ganttproject.gui.UIUtil;

/**
 * Builds standard dialog windows in GanttProject
 *
 * @author dbarashev (Dmitry Barashev)
 */
class DialogBuilder {
  private static class Commiter {
    private boolean isCommited;

    void commit() {
      isCommited = true;
    }

    boolean isCommited() {
      return isCommited;
    }
  }

  private static class NotificationViewImpl implements AnimationView {
    private final JDialog myDlg;
    private BalloonTip myBalloon;
    private Runnable myOnHide;
    private final JButton myNotificationOwner;

    public NotificationViewImpl(JDialog dlg, JButton notificationOwner) {
      myDlg = dlg;
      myNotificationOwner = notificationOwner;
      notificationOwner.setAction(new AbstractAction("Error") {
        @Override
        public void actionPerformed(ActionEvent e) {
          showBalloon();
        }
      });
    }
    protected void showBalloon() {
      myBalloon.setVisible(true);
    }
    @Override
    public boolean isReady() {
      return myDlg.isVisible();
    }

    @Override
    public boolean isVisible() {
      return myBalloon != null && myBalloon.isVisible();
    }

    @Override
    public void setComponent(final JComponent component, JComponent owner, final Runnable onHide) {
      myNotificationOwner.setVisible(true);
      myBalloon = new BalloonTip(myNotificationOwner, component, new EdgedBalloonStyle(Color.WHITE, Color.BLACK),
          BalloonTip.Orientation.LEFT_ABOVE, BalloonTip.AttachLocation.ALIGNED, 30, 10, false);
      myBalloon.setVisible(false);
      myOnHide = onHide;
    }

    @Override
    public void close() {
      if (myBalloon != null) {
        myBalloon.setVisible(false);
      }
      myOnHide.run();
      myNotificationOwner.setVisible(false);
    }
  }

  private static class DialogImpl implements Dialog {
    /** Original animation view, used to set it back when the dialog is closed again */
    private AnimationView myOriginalAnimationView;
    private final JDialog myDlg;
    private final JFrame myMainFrame;
    private final NotificationManager myNotificationManager;
    private JButton myButton;

    DialogImpl(JDialog dlg, JFrame mainFrame, NotificationManager notificationManager) {
      myDlg = dlg;
      myMainFrame = mainFrame;
      myNotificationManager = notificationManager;
    }
    @Override
    public void hide() {
      if (myDlg.isVisible()) {
        myDlg.setVisible(false);
        myDlg.dispose();
      }
      myNotificationManager.setAnimationView(myOriginalAnimationView);
    }

    @Override
    public void show() {
      myOriginalAnimationView = myNotificationManager.setAnimationView(new NotificationViewImpl(myDlg, myButton));
      center(Centering.WINDOW);
      myDlg.setVisible(true);
    }

    @Override
    public void layout() {
      myDlg.pack();
    }

    @Override
    public void center(Centering centering) {
      DialogAligner.center(myDlg, myMainFrame, centering);
    }

    void setNotificationOwner(JButton button) {
      myButton = button;
    }

  }
  private final JFrame myMainFrame;

  DialogBuilder(JFrame mainFrame) {
    myMainFrame = mainFrame;
  }

  /**
   * Creates a dialog given its {@code title}, {@code content} component and an array
   * of actions which are represented as buttons in the bottom of the dialog. Actions
   * which extend {@link OkAction} or {@link CancelAction} will automatically close
   * dialog when invoked.
   *
   * @param content dialog content component
   * @param buttonActions actions for the button row
   * @param title dialog title
   * @return dialog object
   */
  Dialog createDialog(Component content, Action[] buttonActions, String title, final NotificationManager notificationManager) {
    final JDialog dlg = new JDialog(myMainFrame, true);
    final DialogImpl result = new DialogImpl(dlg, myMainFrame, notificationManager);
    dlg.setTitle(title);
    final Commiter commiter = new Commiter();
    Action cancelAction = null;
    JPanel buttonBox = new JPanel(new GridLayout(1, buttonActions.length, 5, 0));
    for (final Action nextAction : buttonActions) {
      JButton nextButton = null;
      if (nextAction instanceof OkAction) {
        final JButton _btn = new JButton();
        final AbstractAction _delegate = (AbstractAction) nextAction;
        OkAction proxy = new OkAction() {
          // These two steps handel the case when focus is somewhere in text input
          // and user hits Ctrl+Enter
          // First we want to move focus to OK button to allow focus listeners, if any,
          // to catch focusLost event
          // Second, we want it to happen before original OkAction runs
          // So we wrap original OkAction into proxy which moves focus and schedules "later" command
          // which call the original action. Between them EDT sends out focusLost events.
          final Runnable myStep2 = new Runnable() {
            @Override
            public void run() {
              result.hide();
              commiter.commit();
              nextAction.actionPerformed(null);
              _delegate.removePropertyChangeListener(myDelegateListener);
            }
          };
          final Runnable myStep1 = new Runnable() {
            @Override
            public void run() {
              _btn.requestFocus();
              SwingUtilities.invokeLater(myStep2);
            }
          };
          @Override
          public void actionPerformed(final ActionEvent e) {
            SwingUtilities.invokeLater(myStep1);
          }
          private void copyValues() {
            for (Object key : _delegate.getKeys()) {
              putValue(key.toString(), _delegate.getValue(key.toString()));
            }
            setEnabled(_delegate.isEnabled());
          }
          private PropertyChangeListener myDelegateListener = new PropertyChangeListener() {
            @Override
            public void propertyChange(PropertyChangeEvent evt) {
              copyValues();
            }
          };
          {
            _delegate.addPropertyChangeListener(myDelegateListener);
            copyValues();
          }
        };
        _btn.setAction(proxy);
        nextButton = _btn;

        if (((OkAction)nextAction).isDefault()) {
          dlg.getRootPane().setDefaultButton(nextButton);
        }

      }
      if (nextAction instanceof CancelAction) {
        cancelAction = nextAction;
        nextButton = new JButton(nextAction);
        nextButton.addActionListener(new ActionListener() {
          @Override
          public void actionPerformed(ActionEvent e) {
            result.hide();
            commiter.commit();
          }
        });
        dlg.getRootPane().getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(
            KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), nextAction.getValue(Action.NAME));
        dlg.getRootPane().getActionMap().put(nextAction.getValue(Action.NAME), new AbstractAction() {
          @Override
          public void actionPerformed(ActionEvent e) {
            nextAction.actionPerformed(e);
            result.hide();
          }
        });
      }
      if (nextButton == null) {
        nextButton = new JButton(nextAction);
      }
      buttonBox.add(nextButton);
      KeyStroke accelerator = (KeyStroke) nextAction.getValue(Action.ACCELERATOR_KEY);
      if (accelerator != null) {
        dlg.getRootPane().getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(accelerator, nextAction);
        dlg.getRootPane().getActionMap().put(nextAction, nextAction);
      }
    }
    dlg.getContentPane().setLayout(new BorderLayout());
    dlg.getContentPane().add(content, BorderLayout.CENTER);

    JButton errorButton = new JButton("Error");
    errorButton.setBackground(UIUtil.ERROR_BACKGROUND);
    //errorLabel.setBorder(BorderFactory.createCompoundBorder(errorLabel.getBorder(), BorderFactory.createEmptyBorder(2,2,2,2)));
    JPanel buttonPanel = new JPanel(new BorderLayout());
    buttonPanel.setBorder(BorderFactory.createEmptyBorder(0, 0, 5, 5));
    buttonPanel.add(buttonBox, BorderLayout.EAST);
    buttonPanel.add(errorButton, BorderLayout.WEST);
    dlg.getContentPane().add(buttonPanel, BorderLayout.SOUTH);
    result.setNotificationOwner(errorButton);
    errorButton.setVisible(false);

    dlg.setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
    final Action localCancelAction = cancelAction;
    dlg.addWindowListener(new WindowAdapter() {
      @Override
      public void windowClosed(WindowEvent e) {
        if (localCancelAction != null && !commiter.isCommited()) {
          localCancelAction.actionPerformed(null);
        }
      }
    });
    dlg.pack();
    return result;
  }
}

File: ganttproject/src/main/java/net/sourceforge/ganttproject/DialogBuilder.kt
/*
 * Copyright 2024 BarD Software s.r.o., Dmitry Barashev.
 *
 * This file is part of GanttProject, an opensource project management tool.
 *
 * GanttProject is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 * GanttProject is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sourceforge.ganttproject

import biz.ganttproject.app.DialogController
import biz.ganttproject.app.dialogFxBuild
import biz.ganttproject.colorFromUiManager
import biz.ganttproject.lib.fx.vbox
import biz.ganttproject.walkTree
import javafx.application.Platform
import javafx.embed.swing.SwingNode
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.scene.control.ButtonBar
import javafx.scene.control.ButtonType
import javafx.scene.control.Dialog
import javafx.scene.input.KeyCombination
import javafx.scene.layout.Background
import javafx.scene.layout.BackgroundFill
import javafx.scene.layout.CornerRadii
import javafx.scene.layout.Priority
import javafx.stage.Screen
import javafx.stage.Stage
import net.sourceforge.ganttproject.action.CancelAction
import net.sourceforge.ganttproject.action.OkAction
import net.sourceforge.ganttproject.gui.UIFacade
import java.awt.BorderLayout
import javax.swing.Action
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.SwingUtilities


class DialogImplSwingInFx(content: JComponent, private val buttonActions: Array<Action>, private val title: String): UIFacade.Dialog {
  private lateinit var controller: DialogController
  lateinit var dialog: Dialog<Unit>
  private var isCommitted = false
  private var contentPane = JPanel(BorderLayout()).also { it.add(content, BorderLayout.CENTER) }
  private val commandsOnShown = mutableListOf<Runnable>()
  private val commandsOnHidden = mutableListOf<Runnable>()

  private fun addButtons(buttonActions: Array<Action>) {
    var cancelAction: Action? = null
    for (action in buttonActions) {
      when {
        action is OkAction -> {
          controller.setupButton(ButtonType.OK) {
            it.text = "${action.getValue(Action.NAME)}"
            it.onAction = EventHandler { SwingUtilities.invokeLater {
              action.actionPerformed(null)
            }}
          }
        }
        action is CancelAction -> {
          controller.setupButton(ButtonType.CANCEL) {
            it.text = "${action.getValue(Action.NAME)}"
          }
        }
        else -> {
          controller.setupButton(ButtonType("${action.getValue(Action.NAME)}", ButtonBar.ButtonData.OTHER)) {
            it.addEventFilter(javafx.event.ActionEvent.ACTION) {
              it.consume()
              SwingUtilities.invokeLater {
                action.actionPerformed(null)
              }
            }
          }
        }
      }
    }

//    for (action in buttonActions) {
//      val nextButton = when {
//        action is OkAction -> JButton().also { _btn ->
//          val _delegate = action as AbstractAction
//          val proxy: OkAction = object : OkAction() {
//            // These two steps handel the case when focus is somewhere in text input
//            // and user hits Ctrl+Enter
//            // First we want to move focus to OK button to allow focus listeners, if any,
//            // to catch focusLost event
//            // Second, we want it to happen before original OkAction runs
//            // So we wrap original OkAction into proxy which moves focus and schedules "later" command
//            // which call the original action. Between them EDT sends out focusLost events.
//            val myStep2: Runnable = Runnable {
//              Platform.runLater { controller.hide() }
//              isCommitted = true
//              action.actionPerformed(null)
//              _delegate.removePropertyChangeListener(myDelegateListener)
//            }
//            val myStep1: Runnable = Runnable {
//              _btn.requestFocus()
//              SwingUtilities.invokeLater(myStep2)
//            }
//
//            override fun actionPerformed(e: ActionEvent) {
//              SwingUtilities.invokeLater(myStep1)
//            }
//
//            private fun copyValues() {
//              for (key in _delegate.keys) {
//                putValue(key.toString(), _delegate.getValue(key.toString()))
//              }
//              isEnabled = _delegate.isEnabled
//            }
//
//            private val myDelegateListener = PropertyChangeListener { copyValues() }
//
//            init {
//              _delegate.addPropertyChangeListener(myDelegateListener)
//              copyValues()
//            }
//          }
//          _btn.action = proxy
//          //++
////          if (action.isDefault) {
////            dlg.rootPane.defaultButton = nextButton
////          }
//
//        }
//
//        action is CancelAction -> {
//          cancelAction = action
//          (action.getValue(GPAction.HAS_DIALOG_BUTTON) as? Boolean)?.let {
//            JButton(action).also {
//              it.addActionListener(ActionListener { e: ActionEvent? ->
//                controller.hide()
//                isCommitted = true
//              })
//            }
//          }
//          //++
////          dlg.rootPane.getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(
////            KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), action.getValue(Action.NAME)
////          )
////          dlg.rootPane.actionMap.put(action.getValue(Action.NAME), object : AbstractAction() {
////            override fun actionPerformed(e: ActionEvent) {
////              action.actionPerformed(e)
////              if (result.isEscCloseEnabled()) {
////                result.hide()
////              }
////            }
////          })
//        }
//
//        else -> JButton(action)
//      }
//      if (nextButton != null) {
//        buttonBox.add(nextButton)
//        buttonCount += 1
//      }
//    }

    cancelAction?.let {cancelAction ->
      controller.onClosed = {
        if (!isCommitted) {
          cancelAction.actionPerformed(null)
        }
      }
    }

//    if (buttonCount > 0) {
//      val buttonPanel = JPanel(BorderLayout())
//      buttonPanel.border = BorderFactory.createEmptyBorder(0, 0, 5, 5)
//      //buttonPanel.add(buttonBox, BorderLayout.EAST)
//      contentPane.add(buttonPanel, BorderLayout.SOUTH)
//    }

  }

  override fun show() {
    Platform.runLater {
      dialog = dialogFxBuild {
        controller = it
        val swingNode = SwingNode()

        it.setContent(vbox {
          add(swingNode, null, Priority.ALWAYS)
        })

        if (buttonActions.isNotEmpty()) {
          addButtons(buttonActions)
        }
        SwingUtilities.invokeLater {
          swingNode.content = contentPane
          Platform.runLater {
            dialog.dialogPane.let {
              it.walkTree { node ->
                if (node is ButtonBar) {
                  node.background = Background(BackgroundFill("Panel.background".colorFromUiManager(), CornerRadii.EMPTY, Insets.EMPTY))
                }
              }
              it.layout()
              it.scene.window.sizeToScene()
            }
          }
        }
      }
      dialog.title = title
      onShown {
        commandsOnShown.forEach { it.run() }
      }
      onClosed { commandsOnHidden.forEach { it.run() } }
      dialog.show()
    }
  }

  override fun hide() = Platform.runLater(controller::hide)

  override fun layout() {
    if (this::dialog.isInitialized) {
      dialog.dialogPane.layout()
    }
  }

  override fun center(centering: UIFacade.Centering) {
    if (this::dialog.isInitialized) {
      val stage = dialog.dialogPane.scene.window as Stage
      centerStage(stage, 600.0, 600.0)
    }
  }

  override fun onShown(onShown: Runnable) {
    if (this::dialog.isInitialized) {
      val prevOnShown = dialog.onShown
      dialog.onShown = EventHandler { evt ->
        prevOnShown?.handle(evt)
        onShown.run()
      }
    } else {
      commandsOnShown += onShown
    }
  }

  override fun onClosed(onClosed: Runnable) {
    if (this::dialog.isInitialized) {
      val prevOnClosed = dialog.onHidden
      dialog.onHidden = EventHandler { evt ->
        prevOnClosed?.handle(evt)
        onClosed.run()
      }
    } else {
      commandsOnHidden += onClosed
    }
  }

  override fun isEscCloseEnabled(): Boolean = dialog.dialogPane.scene.accelerators[KeyCombination.keyCombination("ESC")] != null

  override fun setEscCloseEnabled(value: Boolean) {
    dialog.dialogPane.scene.accelerators[KeyCombination.keyCombination("ESC")] =
      if (value) Runnable { controller.hide() }
      else null
  }

}
fun createDialogFx(content: JComponent, buttonActions: Array<Action>, title: String): UIFacade.Dialog =
  DialogImplSwingInFx(content, buttonActions, title)

private fun centerStage(stage: Stage, width: Double, height: Double) {
  val screenBounds = Screen.getPrimary().getVisualBounds()
  stage.x = (screenBounds.getWidth() - width) / 2
  stage.y = (screenBounds.getHeight() - height) / 2
}